### PR TITLE
add sds client/server example for manual testing

### PR DESCRIPTION
**Pull request by wenchenglu**

### CODE REVIEW

File: Gopkg.lock [hunk 1]
```diff
@@ -1459,6 +1460,6 @@
 [solve-meta]
   analyzer-name = "dep"
   analyzer-version = 1
-  inputs-digest = "64e85d7b234617a9afbc0a95c64d1716f88ea99a3bbf475ab377c8d92f1100e4"
+  inputs-digest = "29d2c4c3d4cd0875ee211ab5e55c87a9554e4fb76b3ecec300dd055fd0014c97"
```
Comment (wattli) [inline Gopkg.lock]: You might want to revert this file, which causes circleci dependency error in presubmit: https://circleci.com/gh/istio/istio/38114?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by wenchenglu** [comment]
done

**Comment by myidpt** [comment]
/lgtm
We could add the server code to this package: https://github.com/istio/istio/tree/master/security/pkg/workload
This package holds libraries for interactions between node agent and workload for delivering key/cert, trust bundles, etc.

**Comment by wenchenglu** [comment]
@myidpt 
good point, will have a follow up cl to refactor this to library in the directory you suggested

**Comment by kyessenov** [comment]
Are there plans to aggregate SDS with regular xDS? Specifically, are there
any consistency issues with secrets not being available for routes?

**Comment by myidpt** [comment]
@wenchenglu That sounds good to me.

**Comment by myidpt** [comment]
@kyessenov I think currently we don't plan to aggregate SDS with xDS. For the consistency issue, I guess we are still implementing the SDS on Envoy. I think it's reasonable to start with having Envoy disable the upstream traffic relying on the key/certs.

**Comment by wenchenglu** [comment]
@kyessenov
This is solely for the bootstrap path to get key/cert locally from a local node agent. I don't expect aggregatingg SDS with xDS in this path. Once key/cert got bootstrapped for the workload, we might be able to aggregate SDS for remote key/cert exchange with other xDS if needed. 

although I am not sure what you meant by "any consistency issues with secrets not being available for routes"?

File: security/cmd/node_agent/sds/server.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	return s
+}
+
+func main() {
+	flag.Parse()
+	var opts []grpc.ServerOption
+	if *tls {
+		creds, err := credentials.NewServerTLSFromFile(*certFile, *keyFile)
+		if err != nil {
+			log.Fatalf("Failed to generate credentials %v", err)
```
Comment (costinm) [inline security/cmd/node_agent/sds/server.go]: Failed to load.

Should it wait for some time ( since certs show up with some delay ) ? Or fail ?

Comment (costinm) [reply security/cmd/node_agent/sds/server.go]: Also what's the use case - isn't it over UDS or localhost ?

File: security/cmd/node_agent/sds/server.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+		VersionInfo: "0",
+		Resources:   resources,
+		TypeUrl:     typeURL,
+	}
+
+	return response, nil
+}
+
+// StreamSecrets is not supported.
+func (s *SDSServer) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecretsServer) error {
```
Comment (costinm) [inline security/cmd/node_agent/sds/server.go]: Did you test this with envoy ? My understanding was that fetch is called by REST ( which is json ), while gRPC is calling stream.

**Comment by costinm** [comment]
For the 'client' side - it would be great to try it out by changing the envoy bootstrap file, so we can get a 
test with the secret watcher disabled and real envoy getting secrets from the node agent. This may solve
quite a few problems, excited to see it moving ahead !

**Comment by kyessenov** [comment]
@myidpt @wenchenglu The consistency issue I have in mind is that if any LDS (or RDS?) resource references a secret that's not delivered through SDS, then those LDS/RDS resources get rejected by envoy. If they don't get rejected, envoy would blackhole the traffic since secrets are not delivered at the same time.

File: security/cmd/node_agent/sds/server.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+func (s *SDSServer) GetTLSCertificate() *auth.TlsCertificate {
+	// TODO: Add implementation. Consider define an interface to support
+	// different implementations that can get certificate from different CA
+	// systems including Istio CA and other CAs.
+	return &auth.TlsCertificate{}
+}
+
+// FetchSecrets fetches the X.509 key/cert for a given workload whose identity
+// can be derived from the UDS path where this call is received.
+func (s *SDSServer) FetchSecrets(ctx context.Context, request *api.DiscoveryRequest) (*api.DiscoveryResponse, error) {
```
Comment (kyessenov) [inline security/cmd/node_agent/sds/server.go]: As an optimization, you can ignore requests that have version info "0".

File: security/cmd/node_agent/sds/server.go (hunk 3 above)
Comment (kyessenov) [reply security/cmd/node_agent/sds/server.go]: Fetch is fine for this use-case. You do need to a gRPC gateway for HTTP mapping.

File: security/cmd/node_agent/sds/server.go (hunk 4 above)
Comment (wenchenglu) [reply security/cmd/node_agent/sds/server.go]: This is a server code mainly for manual test. how to set versioninfo is still to be figured out.

File: security/cmd/node_agent/sds/server.go (hunk 2 above)
Comment (wenchenglu) [reply security/cmd/node_agent/sds/server.go]: yes, it is over UDS. This is a server code for manual test. I'll refactor this PR to a library with UDS support only

**Comment by wenchenglu** [comment]
@kyessenov 
Thanks for bringing up the consistency issue. This PR is mostly for bringing up SDS server on the node agent side. We will have a separate effort to address the client (Envoy) side regarding the LDS/RDS issue. @mangchiandjjoe is working on that. We can have this discussion separately.

**Pull request merged**